jb/itch5/ut_array_based_order_book.cpp

```diff
@@ -0,0 +1,604 @@
+  // .. the best quote should change
+  auto actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(100000));
+  BOOST_CHECK_EQUAL(actual.second, 100);
+  BOOST_CHECK_EQUAL(r, true); // first order
+  // .. and the book_depth should be incremented
+  BOOST_CHECK_EQUAL(tested.count(), 1);
+
+  // ... adding above the best offer has no effect ...
+  r = tested.add_order(price4_t(100100), 300);
+  actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(100000));
+  BOOST_CHECK_EQUAL(actual.second, 100);
+  BOOST_CHECK_EQUAL(r, false);
+  // .. and the book_depth should be incremented
+  BOOST_CHECK_EQUAL(tested.count(), 2);
+
+  // ... update at the offer increases the qty ...
+  r = tested.add_order(price4_t(100000), 400);
+  actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(100000));
+  BOOST_CHECK_EQUAL(actual.second, 500);
+  // handler should return true... it is an inside change
+  BOOST_CHECK_EQUAL(r, true);
+  // .. and the book_depth should not change
+  BOOST_CHECK_EQUAL(tested.count(), 2);
+
+  // ... a better price changes both price and qty ...
+  r = tested.add_order(price4_t(99900), 200);
+  actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(99900));
+  BOOST_CHECK_EQUAL(actual.second, 200);
+  BOOST_CHECK_EQUAL(r, true);
+  // .. and the book_depth should be incremented
+  BOOST_CHECK_EQUAL(tested.count(), 3);
+
+  // ... decrease above the offer has no effect ...
+  r = tested.reduce_order(price4_t(100000), 400);
+  actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(99900));
+  BOOST_CHECK_EQUAL(actual.second, 200);
+  BOOST_CHECK_EQUAL(r, false);
+  // .. and the book_depth should not change
+  BOOST_CHECK_EQUAL(tested.count(), 3);
+
+  // ... even when it is over the existing quantity ...
+  r = tested.reduce_order(price4_t(100000), 200);
+  actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(99900));
+  BOOST_CHECK_EQUAL(actual.second, 200);
+  BOOST_CHECK_EQUAL(r, false);
+  // .. and the book_depth should be incremented
+  BOOST_CHECK_EQUAL(tested.count(), 2);
+
+  // ... deleting the best offer uncovers the best price ...
+  r = tested.reduce_order(price4_t(99900), 200);
+  actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(100100));
+  BOOST_CHECK_EQUAL(actual.second, 300);
+  BOOST_CHECK_EQUAL(r, true);
+  // .. and the book_depth should be incremented
+  BOOST_CHECK_EQUAL(tested.count(), 1);
+
+  // ... deleting the remaining price takes the book depth to 0
+  r = tested.reduce_order(price4_t(100100), 300);
+  actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(200000UL * 10000));
+  BOOST_CHECK_EQUAL(actual.second, 0);
+  // handler should return true
+  BOOST_CHECK_EQUAL(r, true);
+  // .. and the book_depth should be decremented
+  BOOST_CHECK_EQUAL(tested.count(), 0);
+}
+
+/**
+ * @test Verify that the buy side of array_based_order_book works as
```

Comment:
ACK.

I will add better description when moving prices to and from the bottom levels as well